docs/resources/equinix_fabric_routing_protocols.md

```diff
@@ -0,0 +1,230 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "equinix_fabric_routing_protocols Resource - terraform-provider-equinix"
+subcategory: "Fabric"
+description: |-
+  Fabric V4 API compatible resource allows creation and management of Equinix Fabric routing protocols
+  ~> Note Equinix Fabric v4 resources and datasources are currently in Beta. The interfaces related to equinix_fabric_ resources and datasources may change ahead of general availability. Please, do not hesitate to report any problems that you experience by opening a new issue https://github.com/equinix/terraform-provider-equinix/issues/new?assignees=&labels=bug&projects=&template=bug.yml&title=%5BBug%5D%3A+equinix_fabric_routing_protocols
+---
+
+# equinix_fabric_routing_protocols (Resource)
+
+Fabric V4 API compatible resource allows creation and management of Equinix Fabric routing protocols
+
+~> **Note** Equinix Fabric v4 resources and datasources are currently in Beta. The interfaces related to `equinix_fabric_` resources and datasources may change ahead of general availability. Please, do not hesitate to report any problems that you experience by opening a new [issue](https://github.com/equinix/terraform-provider-equinix/issues/new?template=bug.md)
+
+API documentation can be found here - https://developer.equinix.com/dev-docs/fabric/api-reference/fabric-v4-apis#routing-protocols
+
+
+## Example Usage
+
+
+Create Direct and BGP Routing Protocol on Some Connection UUID:
+```hcl
+provider "equinix" {
+  client_id     = "<equinix_client_id>"
+  client_secret = "<equinix_client_secret>"
+}
+
+resource "equinix_fabric_routing_protocols" "routing_protocols" {
+  connection_uuid = "b0eb3892-6404-442b-85c0-4f8fcf53d123"
```

Comment:
I have some questions about this resource that will be easier to answer in terms of this documentation example.

Will a single `equinix_fabric_routing_protocols` explicitly and comprehensively cover all of direct and BGP routing protocols for a specific connection?

If so:
- Is it not sufficient to use the connection id as the `SetId` for this TF resource? Can't we fetch and sync all routes from connection id alone?
- Will any routes not defined in this block be deleted?
- should we provide a warning that only a single routing_protocol resource per connection uuid should be active (count != 0) at a time
- Multiple direct and BGP routes seem to be supported (can we show that in the example)
- Is there a universal documented hard API upper bound on the number of routes? Should this be defined as MaxItems?


If not:
- I wonder if the plural resource name implies otherwise (that it should be a comprehensive resource)
- I assume we are doing so to avoid touching routes not defined in TF?
  - we could surface those routes as smaller resources (one type or setting to differentiate BGP vs direct)
  - we wouldn't be able to use the bulk API endpoint, so perhaps that is the reason we took this path?

In either case:
- How do we use SetId and ImportId in a way that reflects that there may be multiple Direct or BGP sub-resources?

I think it would be helpful to include tests steps (on the existing tests) that:
- add direct and bgp rules from a set
- remove direct and bgp rules from a set
- juggle (move) direct and BGP rules to another set in the same resource (this is what I'm most interested in)
- ^ include an import test on this step (I suspect the additional rules would be dropped today)

Reply:
@displague, only one of Direct and one of BGP are allowed on a connection resource at a time:

https://github.com/equinix-product/interconnection_ic-routing-protocol-management/blob/db8f14a862f7cf9f9deb6283e0d9bc11f119e061/src/main/java/com/equinix/interconnection/fabric/routing/handler/RoutingProtocolHandler.java#L159-L168

This resource is to manage the Single Direct and BGP routing protocol resources under a connection. There will not be more than a single of each. I should add MaxItems = 1 to those set types to reflect that.

This resource is to manage the synchronization between Direct and BGP. Direct cannot be deleted before BGP if both are applied to the connection; an error will be returned if attempted. Additionally, you cannot send a POST for Direct or BGP after they have been created to overwrite; it will return an error saying only one of each type is allowed.

This restriction I believe simplifies your queries. Let me know if that's not the case though.